### PR TITLE
[cmake] Work around a bug in the swift-driver by passing host side libraries from LLVMSupport to linker jobs via a -Xlinker flag.

### PR DESCRIPTION
**Pull request by gottesmm**
Clang and Swift both support this so we can do this until the actual fix lands
in a host toolchain to unblock ErikE.

The specific problem is the swift driver should just push tbd files through to
the linker, but it treats it as an input file. I am going to be putting a fix
into 5.5. This patch in the mean time filters out the tbd files in cmake. This
is a temporary fix.

### CODE REVIEW

File: cmake/modules/SwiftSharedCMakeConfig.cmake [hunk 1]
```diff
@@ -98,7 +98,37 @@ macro(swift_common_standalone_build_config_llvm product)
   endif()
 
   if(LLVM_ENABLE_ZLIB)
-   find_package(ZLIB REQUIRED)
+    find_package(ZLIB REQUIRED)
```
Comment (compnerd) [inline cmake/modules/SwiftSharedCMakeConfig.cmake:101]: Unnecessary whitespace change.

File: cmake/modules/SwiftSharedCMakeConfig.cmake [hunk 2]
```diff
@@ -98,7 +98,37 @@ macro(swift_common_standalone_build_config_llvm product)
...
+      if (TARGET ${lib})
+        list(APPEND new_libraries "-Xlinker;$<TARGET_FILE:${lib}>")
+        continue()
+      endif()
+
+      get_filename_component(LIB_FILENAME_COMPONENT ${lib} LAST_EXT)
+      if ("${LIB_FILENAME_COMPONENT}" STREQUAL ".tbd")
+        # This needs to be an actual string argument since if we use a list,
+        # cmake seems to swallow the -Xlinker. But we need that -Xlinker for our
+        # work around!
```
Comment (compnerd) [inline cmake/modules/SwiftSharedCMakeConfig.cmake:123]: Please use `SHELL:` instead of this (there are examples of this in the Windows path for Swift).

**Review by compnerd — CHANGES_REQUESTED** [review]

File: cmake/modules/SwiftSharedCMakeConfig.cmake (hunk 1 above)
Comment (gottesmm) [reply cmake/modules/SwiftSharedCMakeConfig.cmake:101]: It should be two space. I think its pretty harmless of a change.

File: cmake/modules/SwiftSharedCMakeConfig.cmake (hunk 2 above)
Comment (gottesmm) [reply cmake/modules/SwiftSharedCMakeConfig.cmake:123]: SGTM

Comment (gottesmm) [reply cmake/modules/SwiftSharedCMakeConfig.cmake:123]: For those who are unaware, there is a deduplication step that can swallow options. To work around this, one can use SHELL: as a prefix to the string to work around this when necessary. It is documented here: https://cmake.org/cmake/help/v3.20/command/add_compile_options.html?highlight=shell#arguments but it is a bit hidden from search engines and the like.

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: cmake/modules/SwiftSharedCMakeConfig.cmake (hunk 1 above)
Comment (compnerd) [reply cmake/modules/SwiftSharedCMakeConfig.cmake:101]: Yeah, it is, just a habitual remark on unnecessary changes :) whitespace changes like that dont actually matter to the functionality

**Review by compnerd — APPROVED** [review]
Please add a comment for the three cases being handled.

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Review by drexin — APPROVED** [review]
👍

File: cmake/modules/SwiftSharedCMakeConfig.cmake [hunk 3]
```diff
@@ -98,7 +98,43 @@ macro(swift_common_standalone_build_config_llvm product)
...
+      get_filename_component(LIB_FILENAME_COMPONENT ${lib} LAST_EXT)
+      if ("${LIB_FILENAME_COMPONENT}" STREQUAL ".tbd")
+        list(APPEND new_libraries "SHELL:-Xlinker ${lib}")
+        continue()
+      endif()
+
+      list(APPEND new_libraries "${lib}")
+    endforeach()
+
+    set_target_properties(LLVMSupport PROPERTIES INTERFACE_LINK_LIBRARIES "${new_libraries}")
```
Comment (compnerd) [inline cmake/modules/SwiftSharedCMakeConfig.cmake:137]:
```suggestion
    set_target_properties(LLVMSupport PROPERTIES INTERFACE_LINK_OPTIONS "${new_libraries}")
```

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test
